File: 46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
List = list
class Solution:
    # import math
    # import random
    # def permute(self, nums: List[int]) -> List[List[int]]:
    #     rst = []
    #     while len(rst) != math.factorial(len(nums)):
    #         random.shuffle(nums)
    #         if nums not in rst:
    #             rst.append(nums[:])
    #     return rst
    def permute(self, nums: List[int]) -> List[List[int]]:
        rst = []
        tmp_rst = []
        hash_map = {}
        def recursive():
            if not nums:
                rst.append(tmp_rst[:])
                return None
            for item in nums[:]:
                tmp_rst.append(item)
                nums.remove(item)
                recursive()
                nums.append(item)
                tmp_rst.pop()
        recursive()
        return rst

# @lc code=end

File: 46.全排列.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 */

// @lc code=start
#include<vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> added = vector<bool>(nums.size(), true);
        vector<vector<int>> ret;
        vector<int> tmp;
        generate(added, tmp, ret, nums);
        return ret;
    }

    void generate(vector<bool>& added, vector<int>& tmp, vector<vector<int>>& ret, vector<int>& nums){
        if(added.size() == tmp.size()){
            ret.emplace_back(vector<int>(tmp.begin(), tmp.end()));
            return ;
        }
        for(int i = 0; i < added.size(); i++){
            if(added[i]){
                added[i] = false;
                tmp.emplace_back(nums[i]);
                generate(added, tmp, ret, nums);
                tmp.pop_back();
                added[i] = true;
            }
        }
    }
};
// @lc code=end

